### DISCUSSION TITLE
Two questions about Wan Training : about the FPS, and about the media chunk size.

### DISCUSSION THREAD

**Discussion by dorpxam** [kohya-ss/musubi-tuner#389] [Q&A]
I open this Q&A about two problem I encounter during my training sessions.

Before starting with the questions, I precise that talk about Wan training mainly, but except for the FPS question, the second point (chunk size) can be apply for other models.

About the training settings. I use:
```
--network_module networks.lora_wan
--network_dim 32
--network_alpha 32
--network_args "loraplus_lr_ratio=4" "exclude_patterns=[r'.*single_blocks.*']"
--network_dropout 0.1
--gradient_checkpointing
--gradient_accumulation_steps 2
--optimizer_type came_pytorch.CAME
--optimizer_args "weight_decay=1e-2" "betas=(0.9, 0.999, 0.9999)" "eps=(1e-30, 1e-16)"
--learning_rate 2e-5
--lr_scheduler "constant_with_warmup"
--lr_warmup_steps 100
```
For a Character-LoRA, I use:
```
--timestep_sampling sigmoid
--discrete_flow_shift 1.0
```
For a Concept-LoRA, I use:
```
--timestep_sampling shift
--discrete_flow_shift 7.0
```
I use a `Rank of 32` because I've notice a better quality in particular with regard to the character-LoRA. The final size is `300MB` that, for me, it's a good comprise between size and quality. The CAME optimizer, the LoRA+, the dropout of 10% ... all theses ideas come from @Sarania, thank's for is advice. I just add a 2-steps gradient accumulation that I think is a good point when like me you have a low-VRAM GPU and you don't use batch processing. I've seen that some users use 4-steps gradient accumulation.

Globally, with a good dataset, this configuration is optimum for me. Maybe you've advice? You're welcome to share with us of course.

OK, now let's go for my questions:

## The WAN FPS case:
Most of the time, I prepare my dataset so that it does not require any preprocessing at the caching level. The videos are resized and the framerate modified in Topaz Video, the images are resampled in Photoshop. So regarding the training of the Wan model, the framerate is 16FPS. And this is where the bottom hurts.

Indeed, a 16FPS framerate is nonsense at a time when the videos are in 60FPS or more, and the screens have high refresh rates (here 165Hz). I understand that this is a choice probably related to performance issues at the level of the model itself, but what about the training?

You have already noticed that some LoRA are poorly trained. Certainly with framerate of 24 or 30 FPS which has the effect of a slow-motion type rendering. However, training in 16 FPS poses a real problem when it comes to training very dynamic movements.

I challenge anyone to create a LoRA of the character of Bruce Lee who puts lightning kicks and punches. It's simple, it's just impossible due to the low framerate. The only method that comes to mind would be to train the LoRA with videos at a framerate of say 32 FPS and tell the user that the generation framerate must be divided by two in order to obtain an approximately correct rendering.

By this, the Wan model definitely seems hermetic to fast, really fast movements. What do you think about it? Have you ever faced this problem? Have you found a trick that allows you to work around this problem?

## The learning of long scene during training:
Here I am still digging into the arcana of model training. Have you noticed that some LoRA seem to work poorly when they are trained with short video sequences?

Let's start from the assumption that the default rendering with Wan is 81 frames (5 seconds in 16 FPS + 1 frame). Moreover, in terms of performance during generation, this seems to be the optimum number. Now let's imagine, let's say a training of a concept-LoRA for a liquid that flows on a surface. And let's imagine that the training was done on shorter videos, sequences of 49 frames (3 seconds in 16 FPS + 1 frame). 

By using the LoRA to generate a classic 81-frame video, it's a safe bet that your liquid will disappear from the surface as soon as the video has exceeded 49 frames (ie between 50 and 81 frames). Even if the model generalizes well and produces movements not seen during training, the problem of duration remains.

By training the LoRA with longer videos within a certain limit, the effect can be circumvented but there is no way to know if it will not occur on longer videos that you cannot produce on your machine due to lack of VRAM.

So here is my question. Instead of creating datasets of short sequences that require both preparation work: selection, cutting, resizing, prompting... Wouldn't it be ideal to 'show' the complete video(s) to the model using a single prompt that describes the entire video.

On this subject, I noticed that some users use prompts with time indications like [start:, middle:, end:] to describe the content to be generated, and it seems to work well with Wan.

Technically, the majority of LoRA creators seem to use only the `head` and `chunk` methods for `frame_extraction`, while `musubi-tuner` allows a multitude of variations that can be interesting in this specific case.

OKAY. So let's get to the heart of the matter. Let's imagine a relatively long sequence shot. I wouldn't mention the famous sequence shot of the restaurant in Scorcese's Goodfellas, but let's just imagine a 20-second shot with the camera following the movement of a character on a street, which turns at a street corner and follows the movement of the character who enters a store.

Wan, like many other models, was trained on short sequences and basic camera movements. I would not mention the list of movements that are easily available online, but let's imagine that I want to train a LoRA about character movements in an urban environment that are not limited to a poor catwalk of a sexy bimbo.

Ideally, it is necessary to show the sequences to the model so that he can generalize the concepts. The problem is that it is rarely possible with a local machine to train the model on long-duration sequences, even less at high resolutions. Moreover, I am not sure that the model itself is capable of generalizing over long durations when intrinsically it has been trained on short sequences.

However, `musubi-tuner` offers an interesting possibility with the `slide` or `uniform` methods that allow you to automatically sequence long videos in regular portions with a certain number of overlapping frames. Have any of you already used these methods? In what context precisely?

Here, I appeal to the science of @kohya-ss in order to understand a little better how training works with this type of method? I imagine that fundamentally, it is not much different from the `chunk` or `head` methods, that the sequences are passed to training with the cache of the text encoder for the sequence. But if you have some crunchy details to give about the pure functioning of the model, I'm very interested in that.

In conclusion and to summarize the background of my thought. Can training on long sequences be beneficial with regard to the generalization of concepts and the diversity/quality of the videos generated on these bases? And does segmenting a sequence (without using the `full` method) still allow the model to understand that it is a complete sequence (use of overlapping between sequences)?

Sorry if I was verbose and painful to read. It's important to make yourself understood and I hope that's been the case. Thank you to those who will take the time to respond and share their experience.

**Comment by tavyscrolls** [comment]
I'm still in the experimental stage but I assume prefixing your caption with "Slow motion video of" with reference video at 30 fps on top of some resampled 16 fps videos without the prefix would be a good starting point. For practical purposes, I even wonder about resampling to 16 and extracting every-other-frame with the prefix "Sped up video of" so longer sequences can be fit into context as well as prompting it so the generation can be quickly interpolated to double length (or 4x interpolation with 2x playback speed for a native ~30fps). Might end up needing to increase network dim to 64, idk.

I do think it's good practice to do a mix of higher-res/short samples and lower-res/full length samples and I've also got a medium/medium dataset too. And I captioned every sample twice, so each bucket has 2 captions, one being a sentence or 2 and one being highly detailed. Basically you want it to learn to not depend on any one aspect so doesn't just generalize within a small range of conditions. This might become an issue with a small dataset though, fortunately I've got enough samples that I can make each dataset without duplicates if I have to, especially if I sample from different places of each video.

Btw, does anyone know how Musubi selects the samples for each step? And setting batch size in general vs per-dataset? Not for repeats or dataset balancing, but for training on mixed examples.

**Comment by dorpxam** [comment]
> I'm still in the experimental stage but I assume prefixing your caption with "Slow motion video of" with reference video at 30 fps on top of some resampled 16 fps videos without the prefix would be a good starting point. For practical purposes, I even wonder about resampling to 16 and extracting every-other-frame with the prefix "Sped up video of" so longer sequences can be fit into context as well as prompting it so the generation can be quickly interpolated to double length (or 4x interpolation with 2x playback speed for a native ~30fps). Might end up needing to increase network dim to 64, idk.

The idea of presenting both a slow-motion video and a video with a low frame rate is a good one. Indeed, if the model can learn the link between the two sequences, it can do the trick. I think I'll try it myself to build a karate LoRA with some very fast moves. This is a good exercise and will help clarify things. I'll report progress on this point here as soon as I get around to it.

> I do think it's good practice to do a mix of higher-res/short samples and lower-res/full length samples and I've also got a medium/medium dataset too. And I captioned every sample twice, so each bucket has 2 captions, one being a sentence or 2 and one being highly detailed. Basically you want it to learn to not depend on any one aspect so doesn't just generalize within a small range of conditions. This might become an issue with a small dataset though, fortunately I've got enough samples that I can make each dataset without duplicates if I have to, especially if I sample from different places of each video.

The question is more about extended length videos. Think about a character-LoRA that have some very specific physical details, for example a particular gait, a specific facial movement, or various attributes that make humans unique. Five seconds of video seems like a very short time to capture the subtleties. The problem is that it's clear that the model learns best on small datasets, around ten videos or images.

I'm currently working on a character LoRA based on multiple training phases. To test the concept, I'm using photos and videos of a supermodel. The advantage is that there are plenty of high-quality media available. The procedure I'm considering is as follows:

1. Pre-train a basic LoRA with a few good-quality photos from different angles. Let's say, front view, profile view, head tilted forward, head tilted backward. Close-up views of the face, and upper body views (bust and face).
2. Use this LoRA as a basis for initializing the weights for subsequent training sessions.
3. Create and train multiple LoRAs based on specific photoshoots and/or time periods. This doesn't mean limiting them to overly specific outfits and hairstyles, but the idea is to use both good-quality images and backstage videos for motion capture.
4. Each iteration uses the weights from the previous LoRA to initialize the starting weights for training.
5. Finally, combine all the LoRAs with the Post-Hoc EMA method provided by musubi-tuner, which works well based on my initial tests.

I don't know if this is wise and/or if it will provide any benefit, but I think it's a good idea to be able to capture a large number of character attributes without having to overfit the size of a single dataset.

**Reply by tavyscrolls** [reply]
> Five seconds of video

Ah, I just finished caching my first 81-frame dataset, I've mostly experimented with 1-2 seconds so far

> I'm currently working on a character LoRA based on multiple training phases.

I tried this with Hunyuan to a relative success (though I think Hunyuan lacks too much animation IP to properly build off, so now I'm on Wan) first by training images (exclude= r'.*single_blocks.*', r'.*double_blocks\.[0-9]\..*') then continuing with video on all double/single blocks as well as training the video for motion then images (and with the exclude) and bringing the untrained layers back into the lora. The first route did not accomplish the quality I was hoping for, though I'd call it an improvement. The second route did improve quality greatly but made the animation similar to reaction .gifs or live portraits. I don't remember the exact LR schedule, it was something like 5e-5 then 1e-5, both on cosine. It may have even benefited from a 3rd pass on video-only all blocks at very low LR but at that point I decided Hunyuan wasn't it.

Anyway, there are half a dozen things that could've affected my results so it might just work for you, good luck.

**Comment by dorpxam** [comment]
> Anyway, there are half a dozen things that could've affected my results so it might just work for you, good luck.

Thanks, and thanks for your advice. Obviously, LoRA's training is something that is done through trial and error. It's a pity that AI in general is not deterministic, it would greatly facilitate things and it would save energy. But in a way, there is that magical and joyful side when the results exceed the expected goal. I will try to record the whole process of the described multi-stage method, including the W&B graphs, so I could do a little experiment report here.

Regarding camera movements, yesterday I thought of this idea for training. Just as it is sometimes interesting to blur faces to avoid fixing the model on a particular character, wouldn't it be interesting to try to train the camera movements on videos converted into depth-map. That is to say, purely abstract in form, but describing the sequences by detailing what it really corresponds to. The idea is to force the model to focus on the global movement more than on the details.

Animation is something that gets lost especially when using optimizations to reduce the number of generation steps. Sometimes, despite an explicit prompt, the video remains static overall except for the movements of the people themselves. If you have any ideas about it, I'm interested!

**Comment by HMCIndie** [comment]
Try discrete_flow_shift at 3.0. That has made fast movements easier for the lora to learn at least for me. When I turn that up the movements just go haywire. Any other experiences with discrete flow shift?
